### DISCUSSION TITLE
Add generic runtime specific `Event` variant

### DISCUSSION THREAD

**Discussion by i509VCB** [hecrj/iced#1790] [Ideas]
At the moment there is no way to communicate entirely runtime dependent user interface `Event`s. iced could add a `Runtime` variant that has a `Box<dyn Any>` payload.

The events I would be trying to communicate to the widgets do not belong in the `PlatformSpecific` variant because some of the events I would like to send are from a custom Wayland runtime (using non-standard surface roles) for iced.

Example of API addition:

```diff
pub enum Event {
    /// A keyboard event
    Keyboard(keyboard::Event),

    /// A mouse event
    Mouse(mouse::Event),

    /// A window event
    Window(window::Event),

    /// A touch event
    Touch(touch::Event),

    /// A platform specific event
    PlatformSpecific(PlatformSpecific),
+
+    /// A runtime dependent event.
+    ///
+    /// This should be used to communicate runtime specific events that do not belong in other variants of [`Event`].
+    Runtime(Box<dyn Any>),
}
```

**Comment by hecrj** [comment]
What is the use case?

**Reply by i509VCB** [reply]
The use case is to support custom iced runtimes (like iced-winit). In the case of what I mentioned, a theoretical iced runtime for a session locker (using the https://wayland.app/protocols/ext-session-lock-v1) would need to communicate events to widgets somehow.

A screen locker behaves nothing like a window so some window event variants make no sense. There is also a locked event widgets might want to know about so there needs to be a way to send that, ideally via the Runtime variant. The idea here is to allow custom runtimes to propagate events without iced needing to support every type of runtime.

**Reply by ids1024** [reply]
Currently https://github.com/pop-os/iced includes patches to `iced_native` so that it's `iced_sctk` library can provide events related to Wayland popups, layer shell surfaces, drag and drop, etc.

It would be nice to make this work with unpatched versions of Iced crates, but that would require upstream Iced including these runtime-specific events, or some hack to communicate these events through a separate channel. Or a feature like this.

cc @wash2

**Reply by wash2** [reply]
Ya maybe something like https://github.com/dtolnay/dyn-clone could be used and something similar for PartialEq?

**Reply by wash2** [reply]
Also a similar variant would need to be added to `Action`, but I guess that wouldn't be as tricky because they don't need to be `Clone` or `PartialEq`

**Comment by ids1024** [comment]
Another possibility would by to make it `Event<T: Runtime>` so this could be handled in a cleaner static way, but perhaps that would pollute a lot of things with type generics.

**Reply by wash2** [reply]
Ya I think the last time I tried to do this it kinda did pollute everything with type generics :sweat_smile:
